### PR TITLE
Implemented Solutions to Problems 2 and 3

### PR DESCRIPTION
**Pull request by jeffbulmer**
Hello, 

I have implemented solutions to the second and third parts of the clever-challenge. The second part was done in golang using a simple text parser to parse the ast. The particular structure of this ast is exploited, though the solutions should work for many other asts. 
The third part of the challenge was solved using R to run some basic classification algorithms on the data, primarily for the purposes of variable selection. This was done in R Markdown, so a summary of the whole R segment can be seen as seq.pdf. Once variable selection had been carried out in R, python was used for sequential analysis. This is implemented in seq.py in the seq folder.

### CODE REVIEW

File: seq/seq.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+#plt.figure()
+#groups = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]
+#i = 1
+#for group in groups:
+#    plt.subplot(len(groups), 1, i)
+#    plt.plot(dataset.values[:, group])
+#    plt.title(dataset.columns[group], y=0.5, loc='right')
+#    i+=1
+#plt.show
```
Comment (blaisegarant) [inline seq/seq.py:40]: Why submit that amount of commented code?

File: seq/seq.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+from keras.layers import Flatten
+from keras.layers import LSTM
+from sklearn.preprocessing import MinMaxScaler
+from sklearn.metrics import mean_squared_error
+from sklearn.model_selection import train_test_split
+from sklearn.datasets.samples_generator import make_blobs
+
+
+#columns to be used decided from basic classification in R
+dataset = pandas.read_csv('sample.csv', usecols=[1,2,4,5,6,7,8,9,10,11,12,13,15,16,18,20], engine='python')
```
Comment (blaisegarant) [inline seq/seq.py:24]: In your R section, you removed columns `-c(1,4,24,17,21,23,14,19,22,25,26,27,28,29,30,31)` (it would have been really easier to read if the numbers were in order)
This would translate to `0,3,13,16,18,20,21,22,23,24,25,26,27,28,29,30` in python. Yet, you used columns 13, 16, 18 and 20 in your model.

File: seq.Rmd [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+rf <- randomForest(factor(class)~., data=data.frame(sample[,-1]))
+
+#Now look at the Variable Importance Plot of the Random Forest to determine relevant variables to the model
+varImpPlot(rf)
+```
+
+For comparison, we'll run a logistic regression using only variables above the threshold
+
+```{r}
+simdat <- sample[,-c(1,4,24,17,21,23,14,19,22,25,26,27,28,29,30,31)]
```
Comment (blaisegarant) [inline seq.Rmd:45]: The first column is an ID so of course, it's not relevant by itself but it can be used to link the information here with the one in the `res.csv` file. Which you neither used nor seem to looked at. Any reason?

File: seq/seq.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+dataset_norm = (dataset - dataset.mean()) / (dataset.max() - dataset.min())
+dataset_norm["class"] = dataset["class"]
```
Comment (blaisegarant) [inline seq/seq.py:26]: You normalized the _class_ then set it back to it's original state which makes total sense but you kept the _timestamp_ feature normalized. Why is that?

Since it's a time series, all future timestamp will only go further and further away from your mean.

File: seq/seq.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    #aggregate
+    agg = pandas.concat(cols, axis=1)
+    agg.columns = names
+    #drop NaNs
+    if dropnan:
+        agg.dropna(inplace=True)
+    return agg
+
+scaler = MinMaxScaler(feature_range=(0,1))
+scaled = scaler.fit_transform(dataset_norm.values.astype('float32'))
```
Comment (blaisegarant) [inline seq/seq.py:66]: Why scale them again? And/or why scale them compared to their mean in the first place?

File: seq/seq.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    for i in range(n_in, 0, -1):
+        cols.append(df.shift(i))
+        names += [('var%d(t-%d)' % (j+1, i)) for j in range(n_vars)]
+    #forecast sequence
+    for i in range(0, n_out):
+        cols.append(df.shift(-i))
+        if i == 0:
+            names += [('var%d(t)'%(j+1)) for j in range(n_vars)]
+        else:
+            names += [('var%d(t+%d)'%(j+1)) for j in range(n_vars)]
```
Comment (blaisegarant) [inline seq/seq.py:56]: This doesn't work but you never used n_out>1 therefore you never had the error.

File: seq/seq.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+#groups = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]
+#i = 1
+#for group in groups:
+#    plt.subplot(len(groups), 1, i)
+#    plt.plot(dataset.values[:, group])
+#    plt.title(dataset.columns[group], y=0.5, loc='right')
+#    i+=1
+#plt.show
+
+def series_to_supervised(data, n_in=1, n_out=1, dropnan=True):
```
Comment (blaisegarant) [inline seq/seq.py:42]: This name is not very descriptive of what your function does. And without comments, it's hard to read.

File: seq/seq.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    n_vars =1 if type(data) is list else data.shape[1]
+    df = pandas.DataFrame(data)
+    cols, names = list(), list()
+    #input sequence
+    for i in range(n_in, 0, -1):
```
Comment (blaisegarant) [inline seq/seq.py:47]: Why use a `for` loop if you're not gonna use the function with `n_in` nor `n_out`>1?

File: seq/seq.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    for i in range(0, n_out):
+        cols.append(df.shift(-i))
+        if i == 0:
+            names += [('var%d(t)'%(j+1)) for j in range(n_vars)]
+        else:
+            names += [('var%d(t+%d)'%(j+1)) for j in range(n_vars)]
+    #aggregate
+    agg = pandas.concat(cols, axis=1)
+    agg.columns = names
+    #drop NaNs
```
Comment (blaisegarant) [inline seq/seq.py:60]: I get what you wanted to do, but as a side note, this can be done using `join` in pandas which would make the whole function easier to read.

File: seq/seq.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    if dropnan:
+        agg.dropna(inplace=True)
+    return agg
+
+scaler = MinMaxScaler(feature_range=(0,1))
+scaled = scaler.fit_transform(dataset_norm.values.astype('float32'))
+scaled[:,1] = scaled[:,1].astype('int')
+reframed = series_to_supervised(scaled, 1, 1)
+reframed.drop(reframed.columns[[16,18,19,20,21,22,23,24,25,26,27,28,29,30,31]], axis=1, inplace=True)
```
Comment (blaisegarant) [inline seq/seq.py:69]: You added a bunch of columns to remove them altogether except one? You lost me there

File: seq/seq.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+#split into train and test sets
+train = reframed.values[1:-(math.floor(len(dataset)*0.3)),:]
+test = reframed.values[-(math.floor(len(dataset)*0.3)):,:]
+#split into inputs and outputs
+train_X, train_y = train[:,:-1], train[:,-1]
+test_X, test_y = test[:,:-1], test[:,-1]
+#reshape for 3D input
```
Comment (blaisegarant) [inline seq/seq.py:77]: Since you're essentially feeding only a 2D dataset, I don't see why you reshape it into a 3D matrix. Anyway, all roads lead to Rome with Keras.

File: seq/seq.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+    else:
+        misses = misses+1
+        
+accuracy = hits/len(truth)
+
+print('Measured Accuracy of Model: %.3f' % (accuracy*100))
+print('Evaluated Accuracy of Model: %.3f' % (scores[1]*100) )
+
+
+
```
Comment (blaisegarant) [inline seq/seq.py:116]: In a nutshell, you treated all this like a classic classification system only adding the last event state into your dataset.
In fact, with the present state of the system, you must predict what will be the state of the future events without knowing the features of the next events. 

In the final phases afterwards, you create new events and try to predict their state. There are some issues with how you create them but this is not what was asked.

File: seq/seq.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+Xnew, _ = make_blobs(n_samples=1, centers=3, n_features=train_X.shape[2], random_state=1)
```
Comment (blaisegarant) [inline seq/seq.py:117]: Two things here:

1. `make_blobs` creates gaussian samples while your model is scaled for values between [0-1]
2. You scaled your timestamp therefore any new event cannot have at a timestamp value <1

Again, this is irrelevant since it's not in line with the question.
